### PR TITLE
aws-sdk-cpp: mark as broken on x86_64-darwin

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

See https://github.com/NixOS/nixpkgs/issues/340087

cc @orivej

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
??? This would literally break nix

https://github.com/NixOS/nixpkgs/blob/31247792c61e44f71eb67ce0584a49bab177f869/pkgs/tools/package-management/nix/common.nix#L79

https://github.com/NixOS/nixpkgs/blob/31247792c61e44f71eb67ce0584a49bab177f869/pkgs/tools/package-management/nix/common.nix#L156-L158

**Comment by GaetanLepage** [comment]
> ??? This would literally break nix

What do you mean ?
The package literally does not build right now.
Maybe I am missing something, but, how would this justify not marking a broken package as such ?

**Comment by eclairevoyant** [comment]
Then someone should fix it?

**Comment by GaetanLepage** [comment]
> Then someone should fix it?

Absolutely, but it has remained broken for more than a week now and it is quite annoying to deal with when reviewing other PRs.

**Comment by emilazy** [comment]
Where are you getting a week from? I see one failed build on https://hydra.nixos.org/job/nixpkgs/trunk/aws-sdk-cpp.x86_64-darwin that looks like it might just be an overloaded builder. @K900 already restarted it.

**Comment by GaetanLepage** [comment]
> Where are you getting a week from?

Sorry, this might have been slightly overestimated. But I have encountered this failure several times those past few days.
Of course, before submitting the build failure issue, I have retried the build several times on the darwin builder.

Maybe this package simply does not build succesfully on the remote darwin builder and I had never noticed it because of it being properly cached on `hydra`.
Do you actually think this is a false positive ? I unfortunately do now own an `x86_64-darwin` system to test this.

**Comment by emilazy** [comment]
Sorry, I assumed it was just a Hydra thing and that it hadn’t been reproduced anywhere else.

We can’t really mark this broken but I’ll try to reproduce on my local native `x86_64-darwin`.

**Comment by emilazy** [comment]
FWIW Hydra just uses Rosetta 2 for `x86_64-darwin` so if you have `aarch64-darwin` like the community builder that’s just as good.

**Comment by GaetanLepage** [comment]
> FWIW Hydra just uses Rosetta 2 for `x86_64-darwin` so if you have `aarch64-darwin` like the community builder that’s just as good.

Yes, I know and maybe this failure is simply caused by the _emulation_...



> We can’t really mark this broken but I’ll try to reproduce on my local native `x86_64-darwin`.

Ok thank you. We should definitely double-check.
On a more general note, if some package is actually and verifiably not building, is there some reason to not explicitly mark it as `broken` ?
I ask this, because the time spent at sorting the failures reported by `nixpkgs-review` between actual regressions and false positive seems very unnecessary to me.
Are there some additional conditions that a package needs to meet before being marked as `broken`... despite not building ?

**Comment by reckenrode** [comment]
aws-sdk-cpp is a dependency of Nix. Marking it brokens means that Nix won’t be available on x86_64-darwin.

Marking packages broke should be a last resort. It’s better to fix them if possible.

**Comment by emilazy** [comment]
> On a more general note, if some package is actually and verifiably not building, is there some reason to not explicitly mark it as `broken` ?

If a package this critical is broken, we can’t just mark it broken and move on. We *have* to fix it or else drop support for the platform entirely. If `stdenv` breaks on one of the big platforms, the answer isn’t “mark it broken”, it’s “hey we need to drop everything and fix this right now”. Reverse dependencies of Nix aren’t *quite* that level of critical, but they’re still vital to basic platform support.

So, in other words, this PR basically equates to “casually drop `x86_64-darwin` support”, hence the pushback :) I don’t think you did anything wrong here but it’s not a viable path forward unless we truly exhaust every other option.

(I guess merging this package wouldn’t strictly make anything worse. The channel won’t progress if Nix won’t build, anyway. But we’d need to fix it and revert it ASAP.)

**Comment by GaetanLepage** [comment]
> If a package this critical is broken, we can’t just mark it broken and move on. We _have_ to fix it or else drop support for the platform entirely. If `stdenv` breaks on one of the big platforms, the answer isn’t “mark it broken”, it’s “hey we need to drop everything and fix this right now”. Reverse dependencies of Nix aren’t _quite_ that level of critical, but they’re still vital to basic platform support.
> 
> So, in other words, this PR basically equates to “casually drop `x86_64-darwin` support”, hence the pushback :) I don’t think you did anything wrong here but it’s not a viable path forward unless we truly exhaust every other option.
> 
> (I guess merging this package wouldn’t strictly make anything worse. The channel won’t progress if Nix won’t build, anyway. But we’d need to fix it and revert it ASAP.)

Thank you for the detailed clarification.

- For this package: I didn't know at first that it was somehow linked to `nix` itself. As it only appeared to fail when `nixpkgs-review`-ing some PRs but not others, I thought that it was not as critical. I now perfectly understand that fixing it should be a priority.
- My question was more general and I was thinking of non-critical packages. Basically, from my understanding, "failing but not marked as broken" is an _undesired_/_temporary_ state. A package that does not build should "quickly" either be fixed or marked as broken (along with the relevant `Build failure` issue). There is no advantage in keeping failing packages around, wasting CPU time at trying to build them.
  Is this analysis correct or should failing non-critical packages be left as is in some specific situations ?

**Comment by emilazy** [comment]
Right, I think your second bullet point is correct. That’s why we do Zero Hydra Failures every release cycle to try and fix as many packages as possible or mark them broken if we can’t. Though of course we much prefer the former to the latter when possible.

**Comment by GaetanLepage** [comment]
Thanks for all the insights, I will close this PR and leave the issue opened for now :)

**Comment by sarahec** [comment]
FYI, I ran the aws-sdk-cpp build on x86_64-Darwin and it uses enough memory to start pushing into swap. It's possible that the intermittent failures you saw were OOM errors potentially due to concurrent builds.

**Comment by emilazy** [comment]
The test failures in https://github.com/NixOS/nixpkgs/issues/340087#issuecomment-2334830046 don’t seem like OOM to me. But it feels like a weird test failure to be flaky in the first place, so who knows what’s really going on there.
